chef/cookbooks/monasca/recipes/master.rb

```diff
@@ -0,0 +1,96 @@
+#
+# Copyright 2017 Fujitsu LIMITED
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitation.
+
+package "ansible"
+
+keystone_settings = KeystoneHelper.keystone_settings(node, @cookbook_name)
+
+directory "/opt/fujitsu" do
+  owner "root"
+  group "root"
+  mode "0755"
+end
+
+# TODO: use package "monasca-installer"
+remote_file "/opt/fujitsu/cmm-suse.tgz" do
+  source node[:monasca][:master][:cmm_tarball_url]
+  owner "root"
+  group "root"
+  mode "0644"
+  action :create_if_missing
+end
+
+execute "extract cmm tarball" do
+  command "tar xf cmm-suse.tgz"
+  cwd "/opt/fujitsu/"
+  not_if { Dir.exist?("/opt/fujitsu/monasca-installer") }
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+template "/opt/fujitsu/monasca-installer/credentials.yml" do
+  source "credentials.yml.erb"
+  owner "root"
+  group "root"
+  mode "0644"
+  variables(
+    keystone_settings: keystone_settings
+  )
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+template "/opt/fujitsu/monasca-installer/cmm-hosts" do
+  source "cmm-hosts.erb"
+  owner "root"
+  group "root"
+  mode "0644"
+  variables(
+    # TODO: use all monasca-server hosts
+    monasca_host: MonascaHelper.monasca_hosts(search(:node, "roles:monasca-server"))[0],
+    ssh_user_monasca: "root",
+    offline_resources_host: CrowbarHelper.get_host_for_admin_url(node),
+    ssh_user_offline_host: "root",
+    keystone_host: keystone_settings["public_url_host"],
+    ssh_user_keystone: "root"
+  )
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+template "/opt/fujitsu/monasca-installer/group_vars/all_group" do
+  source "all_group.erb"
+  owner "root"
+  group "root"
+  mode "0644"
+  variables(
+    keystone_settings: keystone_settings
+  )
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+file "/opt/fujitsu/monasca-installer/.installed" do
```

Comment:
Yep, that is not that obvious... I'm even not 100% sure I did right, but it seems to work fine for me...

The whole Ansible run may take 20-30 minutes. So, we want Ansible to run only if there are any changes in configuration (like in the credentials file or - in near future - when the RPM package for the installer gets updated). The .installed file has action set to create_if_missing, so it will trigger the ansible installer only if the file is missing.
Then the file is removed just before the ansible runs, and is re-created after **successful** run. Thanks to that, if ansible installer failed (for example network problems resulting in some timeouts), it will be re-triggered on the next chef-client run. Without this .installed file, ansible would not be re-tried if there were no changes in the configuration.

Needless to say, I have hardly any experience with Chef, so if there is a better way of solving this potential problem, I'll be glad to learn about it,